File: to_do_list.py

class TodoList:
    def __init__(self):
        self.to_do_list = []

    def run(self):
        while True:
            #show menu for user to choose from
            print("To-Do List Menu")
            print("1. View all tasks")
            print("2. Add a task")
            print("3. Complete a task")
            print("4. Delete a task")
            print("5. Clear all tasks")

            #ask user what option they want
            choice = input("Enter your choice (1-5): ")

            #based on what the user chooses, it will run the respective functions
            if choice == "1":
                self.view_tasks()

            elif choice == "2":
                self.add_task()

            elif choice == '3':
                self.complete_task()

            elif choice == "4":
                self.remove_task()
            
            elif choice == '5':
                self.to_do_list.clear()
                print("All tasks cleared")

            else:
                print("Invalid choice. Please enter 1-5.")

    #this function will add a task to to-do list
    def add_task(self):
        #ask the user to input a task
        task = {
            'number': len(self.to_do_list) + 1,
            'name': input("Enter task: "),
            'completed': False,
        }
        #adding the task if it does already exist
        if task not in self.to_do_list:
            self.to_do_list.append(task)
            print(f"Task sucessfully added")
        else: 
            print(f"That task already exists.")

    #this function will allow the user to view the full to-do list
    def view_tasks(self):
        if not self.to_do_list:
            print("There are no tasks yet. Add one to get started.")
            return
        
        #mark task appropriately based on status (✓ = completed, ○ = not completed)
        for task in self.to_do_list:
            status = "✓" if task['completed'] else "○"

            #if task is completed, it will have a strikethrough for the task
            task_name = task['name']
            if task['completed']:
                task_name = f"\033[9m{task_name}\033[0m"  # Strikethrough

            print(f"{status} [{task['number']}] {task_name}")

    #this function will mark a task as complete
    def complete_task(self):
        self.view_tasks()
        task_number = int(input(f"Enter task number completed: ")) 
        for task in self.to_do_list:
            if task['number'] == task_number:
                task['completed'] = True
                print(f"✓ Task {task_number} marked as complete!")
                return
            
        print((f"Task {task_number} not found."))

    #this function will remove a task from the to-do list
    def remove_task(self):
        self.view_tasks()
        try:
            task_number = int(input("Enter task number to remove: "))
            for i, task in enumerate(self.to_do_list):
                if task['number'] == task_number:
                    removed = self.to_do_list.pop(i)
                    print(f"Task Sucessfully Removed")
                    return
            print(f"Task {task_number} not found.") 
        except ValueError:
            print(f"Please enter a valid task number.")   
    
#runs the program
if __name__ == "__main__":
    todo_list = TodoList()
    todo_list.run()

File: to_do_list.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

public class Task
{
    public int Id { get; set; }
    public string Description { get; set; }
    public bool Completed { get; set; }
    public string CreatedAt { get; set; }
}

public class TodoList
{
    private List<Task> tasks;
    private const string JsonFilePath = "todos.json";

    public TodoList()
    {
        tasks = new List<Task>();
        LoadTasks();
    }

    private void LoadTasks()
    {
        if (File.Exists(JsonFilePath))
        {
            try
            {
                string jsonString = File.ReadAllText(JsonFilePath);
                tasks = JsonSerializer.Deserialize<List<Task>>(jsonString) ?? new List<Task>();
                Console.WriteLine("Tasks loaded successfully.");
            }
            catch (Exception e)
            {
                Console.WriteLine($"Error loading tasks: {e.Message}");
                tasks = new List<Task>();
            }
        }
    }

    private void SaveTasks()
    {
        try
        {
            var options = new JsonSerializerOptions { WriteIndented = true };
            string jsonString = JsonSerializer.Serialize(tasks, options);
            File.WriteAllText(JsonFilePath, jsonString);
        }
        catch (Exception e)
        {
            Console.WriteLine($"Error saving tasks: {e.Message}");
        }
    }

    public void AddTask()
    {
        Console.Write("Enter task description: ");
        string description = Console.ReadLine();

        if (string.IsNullOrWhiteSpace(description))
        {
            Console.WriteLine("Task description cannot be empty.");
            return;
        }

        int taskId = tasks.Count > 0 ? tasks.Max(t => t.Id) + 1 : 1;
        var newTask = new Task
        {
            Id = taskId,
            Description = description,
            Completed = false,
            CreatedAt = DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss")
        };

        tasks.Add(newTask);
        SaveTasks();
        Console.WriteLine($"Task '{description}' added successfully!");
    }

    public void ViewTasks()
    {
        if (tasks.Count == 0)
        {
            Console.WriteLine("\nNo tasks in your to-do list.");
            return;
        }

        Console.WriteLine("\n=== Your To-Do List ===");
        foreach (var task in tasks)
        {
            string status = task.Completed ? "✓" : "○";
            string description = task.Completed ? $"[DONE] {task.Description}" : task.Description;
            Console.WriteLine($"{task.Id}. {status} {description} (Created: {task.CreatedAt})");
        }
    }

    public void CompleteTask()
    {
        Console.Write("Enter task ID to mark as complete: ");
        if (int.TryParse(Console.ReadLine(), out int taskId))
        {
            var task = tasks.FirstOrDefault(t => t.Id == taskId);
            if (task != null)
            {
                task.Completed = true;
                SaveTasks();
                Console.WriteLine($"Task '{task.Description}' marked as complete!");
            }
            else
            {
                Console.WriteLine("Task not found.");
            }
        }
        else
        {
            Console.WriteLine("Invalid input. Please enter a number.");
        }
    }

    public void RemoveTask()
    {
        Console.Write("Enter task ID to remove: ");
        if (int.TryParse(Console.ReadLine(), out int taskId))
        {
            var task = tasks.FirstOrDefault(t => t.Id == taskId);
            if (task != null)
            {
                tasks.Remove(task);
                RenumberTasks();
                SaveTasks();
                Console.WriteLine($"Task '{task.Description}' removed successfully!");
            }
            else
            {
                Console.WriteLine("Task not found.");
            }
        }
        else
        {
            Console.WriteLine("Invalid input. Please enter a number.");
        }
    }

    public void ClearCompletedTasks()
    {
        int count = tasks.RemoveAll(t => t.Completed);
        if (count > 0)
        {
            RenumberTasks();
            SaveTasks();
            Console.WriteLine($"{count} completed task(s) cleared!");
        }
        else
        {
            Console.WriteLine("No completed tasks to clear.");
        }
    }

    public void GetTaskStatus()
    {
        Console.Write("Enter task ID to check status: ");
        if (int.TryParse(Console.ReadLine(), out int taskId))
        {
            var task = tasks.FirstOrDefault(t => t.Id == taskId);
            if (task != null)
            {
                string status = task.Completed ? "Completed (True)" : "Not Completed (False)";
                Console.WriteLine($"\nTask ID {taskId}: {task.Description}");
                Console.WriteLine($"Status: {status}");
            }
            else
            {
                Console.WriteLine("Task not found.");
            }
        }
        else
        {
            Console.WriteLine("Invalid input. Please enter a number.");
        }
    }

    public void GetTaskAtIndex()
    {
        Console.Write("Enter index (0-based): ");
        if (int.TryParse(Console.ReadLine(), out int index))
        {
            if (index >= 0 && index < tasks.Count)
            {
                var task = tasks[index];
                Console.WriteLine($"\n--- Task at Index {index} ---");
                Console.WriteLine($"ID: {task.Id}");
                Console.WriteLine($"Description: {task.Description}");
                Console.WriteLine($"Completed: {task.Completed}");
                Console.WriteLine($"Created At: {task.CreatedAt}");
            }
            else
            {
                Console.WriteLine($"Index out of range. Valid range: 0-{tasks.Count - 1}");
            }
        }
        else
        {
            Console.WriteLine("Invalid input. Please enter a number.");
        }
    }

    private void RenumberTasks()
    {
        for (int i = 0; i < tasks.Count; i++)
        {
            tasks[i].Id = i + 1;
        }
    }

    public void Run()
    {
        while (true)
        {
            Console.WriteLine("\n=== To-Do List Menu ===");
            Console.WriteLine("1. Add Task");
            Console.WriteLine("2. View Tasks");
            Console.WriteLine("3. Mark Task as Complete");
            Console.WriteLine("4. Remove Task");
            Console.WriteLine("5. Clear Completed Tasks");
            Console.WriteLine("6. Check Task Status");
            Console.WriteLine("7. Get Task at Index");
            Console.WriteLine("8. Exit");
            Console.Write("Choose an option: ");

            string choice = Console.ReadLine();

            switch (choice)
            {
                case "1":
                    AddTask();
                    break;
                case "2":
                    ViewTasks();
                    break;
                case "3":
                    CompleteTask();
                    break;
                case "4":
                    RemoveTask();
                    break;
                case "5":
                    ClearCompletedTasks();
                    break;
                case "6":
                    GetTaskStatus();
                    break;
                case "7":
                    GetTaskAtIndex();
                    break;
                case "8":
                    tasks.Clear();
                    SaveTasks();
                    Console.WriteLine("All tasks cleared. Goodbye!");
                    return;
                default:
                    Console.WriteLine("Invalid choice. Please try again.");
                    break;
            }
        }
    }
}

public class Program
{
    public static void Main(string[] args)
    {
        var todoList = new TodoList();
        todoList.Run();
    }
}